src/testing/vortex/workload.zig

```diff
@@ -0,0 +1,432 @@
+    const results = receive(driver, operation, buffer) catch |err| {
+        switch (err) {
+            error.EndOfStream => return null,
+            else => return err,
+        }
+    };
+    return @unionInit(Result, @tagName(command_tag), results);
+}
+
+fn reconcile(result: Result, command: *const Command, model: *Model) !void {
+    switch (result) {
+        .create_accounts => |entries| {
+            const accounts_new = command.create_accounts;
+
+            // Track results for all new accounts, assuming `.ok` if response from driver is
+            // omitted.
+            var accounts_created: [accounts_count_max]struct {
+                tb.CreateAccountResult,
+                tb.Account,
+            } = undefined;
+            for (
+                accounts_new,
+                accounts_created[0..accounts_new.len],
+            ) |account_new, *account_created| {
+                account_created[0] = .ok;
+                account_created[1] = account_new;
+            }
+
+            for (entries) |entry| {
+                accounts_created[entry.index][0] = entry.result;
+            }
+
+            for (accounts_created[0..accounts_new.len], 0..) |account_created, i| {
+                if (account_created[0] == .ok) {
+                    model.accounts.appendAssumeCapacity(account_created[1]);
+                } else {
+                    log.err("enum: {d}", .{@intFromEnum(account_created[0])});
+                    log.err("got result {s} for event {d}: {any}", .{
+                        @tagName(account_created[0]),
+                        i,
+                        account_created[1],
+                    });
+                    return error.TestFailed;
+                }
+            }
+        },
+        .create_transfers => |entries| {
+            const transfers = command.create_transfers;
+
+            for (entries, 0..) |entry, e| {
+                const transfer = transfers[entry.index];
+
+                // Check that linked transfers fail together.
+                if (entry.index > 0) {
+                    const preceeding_transfer = transfers[entry.index - 1];
+                    if (preceeding_transfer.flags.linked) {
+                        try testing.expect(e > 0);
+                        const preceeding_entry = entries[e - 1];
+                        try testing.expect(preceeding_entry.index == entry.index - 1);
+                        try testing.expect(preceeding_entry.result != .ok);
+                    }
+                }
+
+                // No further validation needed for failed transfers.
+                if (entry.result != .ok) {
+                    continue;
+                }
+
+                if (transfer.flags.pending) {
+                    try testing.expect(!model.pending_transfers.contains(transfer.id));
+                    try model.pending_transfers.put(transfer.id, {});
+                    // We can't let this grow unboundedly:
```

Comment:
Ah, so if we assert this here, there's no reason _not_ to use putAssumingCapacity!